feat(spans): initial MongoDB description scrubbing support

mjq opened the pull request:
We would like to support MongoDB queries in the Queries insight module (which is currently SQL-only).

Metrics will be shown on the Queries insights page when they fulfill the query `has:span.description span.module:db`. Without the changes in this PR, MongoDB spans would either lack a `span.description` or `span.module` so they were hidden from the module.

In addition, to fully support the Queries page, the `span.action` and `span.domain` tags are also required. Without this PR, MongoDB spans' `span.action` was being set if it was present as `db.operation` in span `data`, but `span.domain` was not set.

This PR:

- Allows MongoDB spans to be tagged with `span.module`: `db`
- Adds MongoDB query scrubbing and writes the scrubbed query to `span.description`, conditional on the `organizations:performance-queries-mongodb-extraction` feature flag (added [here](https://github.com/getsentry/sentry/pull/74498))
- Writes the collection name to `span.domain` for MongoDB spans

Because the new query scrubbing is behind a feature flag, orgs without it will continue to have no `span.description` and therefore continue to not show up in the Queries module.

The short-term plan is to turn this on for a single test org to iterate on the frontend experience, and also begin to test the cardinality of the query scrubbing.

The scrubbing keeps the operation/collection key/value pair but otherwise replaces leaf values in the query JSON with `"?"`. This might need to be tuned further if there are cases where people are likely to use high-cardinality keys.

Valid MongoDB spans must:
- contain a JSON-encoded query in their `description`
- have a `db.system` of `"mongodb"`
- have either `db.collection.name` or `db.mongodb.collection` set
- have `db.operation` set

Our JavaScript v8 SDK fulfills these criteria for MongoDB spans, and I believe the latest Python does too (but haven't personally checked).

mjq committed:
feat(spans): initial MongoDB description scrubbing support

We would like to support MongoDB queries in the Queries insight module (which
is currently SQL-only).

Metrics will be shown on the Queries insights page when they fulfill the query
`has:span.description span.module:db`. Without the changes in this PR, MongoDB
spans would either lack a `span.description` or `span.module` so they were
hidden from the module.

In addition, to fully support the Queries page, the `span.action` and
`span.domain` tags are also required. Without this PR, MongoDB spans'
`span.action` was being set if it was present as `db.operation` in span `data`,
but `span.domain` was not set.

This PR:

- Allows MongoDB spans to be tagged with `span.module`: `db`
- Adds MongoDB query scrubbing and writes the scrubbed query to
  `span.description`, conditional on the
`organizations:performance-queries-mongodb-extraction` feature flag
- Writes the collection name to `span.domain` for MongoDB spans

Because the new query scrubbing is behind a feature flag, orgs without it will
continue to have no `span.description` and therefore continue to not show up in
the Queries module.

The short-term plan is to turn this on for a single test org to iterate on the
frontend experience, and also begin to test the cardinality of the query
scrubbing.

The scrubbing keeps the operation/collection key/value pair but otherwise
replaces leaf values in the query JSON with `"?"`. This might need to be tuned
further if there are cases where people are likely to use high-cardinality
keys.

Valid MongoDB spans must:
- contain a JSON-encoded query in their `description`
- have a `db.system` of `"mongodb"`
- have either `db.collection.name` or `db.mongodb.collection` set
- have `db.operation` set

Our JavaScript v8 SDK fulfills these criteria for MongoDB spans.

Co-authored-by: Ash <0Calories@users.noreply.github.com>

mjq committed:
add changelog

mjq commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -33,12 +34,22 @@ const MAX_EXTENSION_LENGTH: usize = 10;
+    Disabled,
+    /// Enable scrubbing of MongoDB span descriptions.
+    Enabled,
+}
```
Because this value has to get passed through so many different functions, it seemed clearer to use an explicit enum argument instead of a dangling boolean for the arguments.

Dav1dde commented on relay-dynamic-config/src/feature.rs:
```diff
@@ -99,6 +99,13 @@ pub enum Feature {
+    ///
+    /// Serialized as `organizations:performance-queries-mongodb-extraction`.
+    #[serde(rename = "organizations:performance-queries-mongodb-extraction")]
+    ScrubMongoDBDescriptions,
```
Does this feature make sense the way it is right now? 

It seems like `mongodb` is unconditionally enabled with this PR only the scrubbing is conditional. For me this seems like we should have a feature flag to enable/disable mongodb alltogether not just the scrubbing which is supposed to help us with cardinality.

mjq replied:
Here's my understanding.

There are two places that Relay currently special-cases MongoDB. The first is disabled databases.

https://github.com/getsentry/relay/blob/70c15da848216cf5a20062393d47d1d10341323f/relay-dynamic-config/src/defaults.rs#L11-L17
https://github.com/getsentry/relay/blob/70c15da848216cf5a20062393d47d1d10341323f/relay-dynamic-config/src/defaults.rs#L123-L126

`is_db` is used to control the presence of certain metrics and tags in [`hardcoded_span_metrics`](https://github.com/getsentry/relay/blob/70c15da848216cf5a20062393d47d1d10341323f/relay-dynamic-config/src/defaults.rs#L120).

This PR removes that check, which means we will produce more metrics. However, note that:
- a number of SDKs (including v8 of the JS SDK) don't actually have `mongodb` as part of the span `op`, and were never subject to this check
- high cardinality of the resulting metrics comes from the `span.description` tag, which is protected against separately (see following)

The second existing special case is in the span description scrubbing.

https://github.com/getsentry/relay/blob/70c15da848216cf5a20062393d47d1d10341323f/relay-event-normalization/src/normalize/span/description/mod.rs#L63-L70

We check again for `mongodb` in the op (which as I mentioned, is often already bypassed), *but* the `is_sql_mongodb` function catches them all at this point by [checking for a `db.system` of `mongodb` or JSON-looking description](https://github.com/getsentry/relay/blob/70c15da848216cf5a20062393d47d1d10341323f/relay-event-normalization/src/normalize/span/description/mod.rs#L146-L152). This is the handling we fall through to when the feature is off.

---

So:
- This PR will change the situation from "some MongoDB spans are counted as `is_db`" to "_all_ MongoDB spans are counted as `is_db`", but
- There will be no increase in cardinality, as all MongoDB span descriptions will continue to be `None` in cases where the flag is not set.

In an ideal world, I would have kept the behaviour exactly the same and only conditionally changed `DISABLED_DATABASES`. But, looking at how `hardcoded_span_metrics`, I can't see any way to get feature flags down to it. So I reasoned that the outcomes above were probably acceptable, but I could definitely be wrong! Please let me know what you think. Thanks so much 🙏

Dav1dde replied:
> a number of SDKs (including v8 of the JS SDK) don't actually have mongodb as part of the span op, and were never subject to this check

I'd be worried about the Otel integrations here, not the SDKs necessarily.

> high cardinality of the resulting metrics comes from the span.description tag, which is protected against separately (see following)

If we have some confidence in the scrubbing (which seems pretty sound to me, we can even start with a lower 'recursion limit'), I'd not use the feature flag at all. You will only find the outliers after it's too late anyways. And the feature flag just adds more conditionals and boilerplate while only toggling scrubbing not the full extraction.
That's the same approach that was taken for Redis.

Dav1dde commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,43 @@ fn scrub_function(string: &str) -> Option<String> {
+    };
+
+    let mut scrubbed_map: Map<String, Value> = Map::new();
+    original.as_object()?.iter().for_each(|(key, value)| {
```
`for_each` is a bit of a [rust anti pattern](https://rust-lang.github.io/rust-clippy/master/index.html#/needless_for_each), you can just use a `for` here.

Dav1dde commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,43 @@ fn scrub_function(string: &str) -> Option<String> {
+        Value::Object(map) => {
+            let mut scrubbed_map = Map::new();
+            map.iter().for_each(|(key, value)| {
+                scrubbed_map.insert(key.clone(), scrub_mongodb_leaf_nodes(value));
```
Do we need a max recursion limit/depth here?

mjq replied:
Added a conservative limit (too deep seems like it would invite cardinality issues).

Dav1dde commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,43 @@ fn scrub_function(string: &str) -> Option<String> {
+
+    let mut scrubbed_map: Map<String, Value> = Map::new();
+    original.as_object()?.iter().for_each(|(key, value)| {
+        scrubbed_map.insert(key.clone(), scrub_mongodb_leaf_nodes(value));
```
I think none of these clones are required, you can either move the elements into the new map or modify the map in place. The in-place modifications seems to be the more efficient and easier solution to this.

Dav1dde commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,43 @@ fn scrub_function(string: &str) -> Option<String> {
+            });
+            Value::Array(scrubbed_arr.to_owned())
+        }
+        _ => Value::String("?".to_string()),
```
```suggestion
        _ => Value::String("?".to_owned()),
```

Dav1dde commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,43 @@ fn scrub_function(string: &str) -> Option<String> {
+            Value::Object(scrubbed_map.to_owned())
+        }
+        Value::Array(arr) => {
+            let mut scrubbed_arr = Vec::new();
```
This can also just be in an inplace modification.

mjq committed:
address review feedback

- replaced copies with in-place modification
- added a recursion limit to the query scrubbing
- used for loops instead of `for_each`
- `to_owned` instead of `to_string`

Dav1dde committed:
eliminate clones from scrubbing

Dav1dde reviewed (dismissed):
As noted I'd consider dropping the feature flag all together, especially since you expect almost no mongodb data in the first place. Just removes a few more conditionals and overhead. Up to you.

Conditionally changing `is_db` would be preferrable and possible, but that requires some deeper changes in Relay we can always do later and I don't want to block you on this.

mjq committed:
Merge branch 'master' into mjq-ash-mongodb-queries

mjq committed:
ignore arg count lint

mjq committed:
Merge branch 'master' into mjq-ash-mongodb-queries

jjbayer commented on relay-dynamic-config/src/feature.rs:
```diff
@@ -99,6 +99,13 @@ pub enum Feature {
+    ///
+    /// Serialized as `organizations:performance-queries-mongodb-extraction`.
+    #[serde(rename = "organizations:performance-queries-mongodb-extraction")]
+    ScrubMongoDBDescriptions,
```
nit: 
```suggestion
    ScrubMongoDbDescriptions,
```

https://rust-lang.github.io/api-guidelines/naming.html#:~:text=In%20UpperCamelCase%2C%20acronyms%20and%20contractions%20of%20compound%20words%20count%20as%20one%20word%3A%20use%20Uuid%20rather%20than%20UUID

jjbayer commented on relay-dynamic-config/src/defaults.rs:
```diff
@@ -121,9 +115,9 @@ pub fn hardcoded_span_metrics() -> Vec<(GroupKey, Vec<MetricSpec>, Vec<TagMappin
-            | RuleCondition::glob("span.op", DISABLED_DATABASES)
-            | RuleCondition::glob("span.description", MONGODB_QUERIES));
+        & !RuleCondition::glob("span.op", DISABLED_DATABASES)
+        & (RuleCondition::eq("span.system", "mongodb")
+            | !RuleCondition::glob("span.description", MONGODB_QUERIES));
```
Took me a while to parse this condition, might benefit from a comment along the lines of "we disallow mongodb, unless `span.system` is set to `mongodb`".

jjbayer commented on relay-event-normalization/src/event.rs:
```diff
@@ -332,6 +337,7 @@ fn normalize(event: &mut Event, meta: &mut Meta, config: &NormalizationConfig) {
             event,
             config.max_tag_value_length,
             config.span_allowed_hosts,
+            config.scrub_mongo_description.clone(),
```
You can derive the [`Copy`](https://doc.rust-lang.org/stable/std/marker/trait.Copy.html) trait on `ScrubMongoDescription` to make the `.clone()` unnecessary.

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -33,12 +34,22 @@ const MAX_EXTENSION_LENGTH: usize = 10;
 /// Domain names that are preserved during scrubbing
 const DOMAIN_ALLOW_LIST: &[&str] = &["localhost"];
 
+#[derive(PartialEq, Clone, Debug)]
```
```suggestion
#[derive(PartialEq, Clone, Copy, Debug)]
```

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -60,6 +71,22 @@ pub(crate) fn scrub_span_description(
+                    .and_then(|data| data.db_collection_name.value())
+                    .and_then(|collection| collection.as_str());
+
+                command.zip(collection).and_then(|(command, collection)| {
```
```suggestion
                if let (Some(command), Some(collection)) = (command, collection) {
```

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +559,50 @@ fn scrub_function(string: &str) -> Option<String> {
+    for value in root.values_mut() {
+        scrub_mongodb_visit_node(value, 3);
+    }
+    root.insert(command, Value::String(collection));
```
nit: `query` might not parse into a valid object, so I would pass `&str` into this function and only convert `command` and `collection` to `String` in this line.

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -1227,9 +1397,119 @@ mod tests {
+
+    mongodb_scrubbing_test!(
+        mongodb_basic_query,
+        "{\"find\": \"documents\", \"showRecordId\": true}",
```
```suggestion
        r#"{"find": "documents", "showRecordId": true}"#,
```

etc,

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -60,6 +71,22 @@ pub(crate) fn scrub_span_description(
+                    .and_then(|command| command.as_str());
+
+                let collection = data
+                    .and_then(|data| data.db_collection_name.value())
```
With SQL, we learned the hard way that collections / tables can contain all kinds of identifiers (see [test cases here](https://github.com/getsentry/relay/blob/b0ddc97eddcf0e3ed173620668e5cb6110ddf422/relay-event-normalization/src/normalize/span/description/sql/mod.rs#L911-L957)). IMO we should pass the mongodb collection name through the same scrubbing as SQL table names to catch those identifers, i.e. apply the [`TABLE_NAME_REGEX`](https://github.com/getsentry/relay/blob/b0ddc97eddcf0e3ed173620668e5cb6110ddf422/relay-event-normalization/src/normalize/span/description/sql/parser.rs#L23-L24).

jjbayer reviewed (commented):
See comment about scrubbing collection names. The rest of the PR looks good to me!

mjq committed:
Merge branch 'master' into mjq-ash-mongodb-queries

mjq committed:
address review feedback

- Rename `ScrubMongoDBDescriptions` to `ScrubMongoDbDescriptions`
- comment complicated is_db conditional
- add `Copy` trait to `ScrubMongoDescription`
- clarify option checking (zip -> conditional)
- move `String` conversion after failure cases
- use r# strings to make embedded JSON easier to read
- scrub collection names from descriptions and tags

jjbayer commented on relay-event-normalization/src/normalize/span/description/mod.rs:
```diff
@@ -532,6 +562,57 @@ fn scrub_function(string: &str) -> Option<String> {
+        scrub_mongodb_visit_node(value, 3);
+    }
+
+    let scrubbed_collection_name =
+        if let Cow::Owned(s) = TABLE_NAME_REGEX.replace_all(collection, "{%s}") {
+            s
+        } else {
+            collection.to_owned()
+        };
```
nit: I _think_ this is the same

```suggestion
    let scrubbed_collection_name = TABLE_NAME_REGEX.replace_all(collection, "{%s}").to_owned();
```

mjq replied:
The above returns the type error:

``` error[E0308]: mismatched types
   --> relay-event-normalization/src/normalize/span/description/mod.rs:575:51
    |
575 |     root.insert(command.to_owned(), Value::String(scrubbed_collection_name));
    |                                     ------------- ^^^^^^^^^^^^^^^^^^^^^^^^- help: try using a conversion method: `.to_string()`
    |                                     |             |
    |                                     |             expected `String`, found `Cow<'_, str>`
    |                                     arguments to this enum variant are incorrect
    |
    = note: expected struct `std::string::String`
                 found enum `std::borrow::Cow<'_, str>`
```

But
```rust
let scrubbed_collection_name = TABLE_NAME_REGEX.replace_all(collection, "{%s}").to_string();
```

compiles, if that seems reasonable?

Either way I'm going to merge without this change just in case, and loop back afterwards.

jjbayer replied:
Yep, that works!

jjbayer commented on relay-event-normalization/src/normalize/span/tag_extraction.rs:
```diff
@@ -626,9 +634,28 @@ pub fn extract_tags(
+                    .and_then(|data| data.db_collection_name.value())
+                    .and_then(|db_collection| db_collection.as_str())
+                    .map(|db_collection| {
+                        if let Cow::Owned(s) = TABLE_NAME_REGEX.replace_all(db_collection, "{%s}") {
+                            s
+                        } else {
+                            db_collection.to_owned()
+                        }
+                    })
```
Same comment about `to_owned`.

It's a bit unfortunate that we scrub the collection name again here, but that's a basic flaw of `extract_tags` (it works on each tag separately instead of collecting all tags for a given span type at once).

mjq committed:
Merge branch 'master' into mjq-ash-mongodb-queries
